File: CPP/ByTier/L/1776_L_Controllers.cpp
// Problem: CF 1776 L - Controllers
// https://codeforces.com/contest/1776/problem/L

/*
 * Problem: Game with Controllers
 * 
 * Purpose:
 *   Given a sequence of + and - symbols, determine if it's possible to achieve a score of 0
 *   by choosing which button to press in each round. Each controller has two buttons with values a and b.
 *   Pressing button 'a' increases the score by 'a' for '+' and decreases it by 'a' for '-',
 *   similar for button 'b'.
 *
 * Algorithm:
 *   - Preprocess the number of '+' signs (f) and '-' signs.
 *   - For each query (controller), consider two cases:
 *     1. If a == b, we can only use one type of button, so check if f == n - f (equal number of + and -)
 *     2. If a != b, simulate the effect: Let x < y be the smaller and larger button.
 *        - We can make at most f buttons of type 'x' and (n-f) buttons of type 'y'.
 *        - For valid solution, there exists an assignment such that total score = 0.
 *        - If we select k buttons of value x, then (f - k) buttons of value y to balance,
 *          the condition is: k * (x - y) + (n - f) * y = 0 => k = (n - f) * y / (y - x)
 *        - Check if this k is a valid number of buttons (between 0 and f).
 *
 * Time Complexity:
 *   - Preprocessing: O(n)
 *   - Queries: O(q), where q is the number of controllers
 *   - Overall: O(n + q)
 *
 * Space Complexity:
 *   - O(n) for storing the input string
 *   - O(1) additional space
 */

#include <stdio.h>
#include <algorithm>
#include <utility>

using namespace std;
typedef long long ll;
char num[200010]; // Stores the sequence of rounds (1-indexed)

int main() {
  int n;
  scanf("%d%s", &n, num + 1); // Read n and string s (1-indexed)
  
  int t, f = 0; // f counts number of '+' characters
  for (int i = 1; i <= n; i++)
    f += (num[i] == '+'); // Count '+' signs

  scanf("%d", &t); // Read number of controllers
  f = min(f, n - f); // Minimize the number of + or - to reduce work (since we're checking for balance)

  while (t--) {
    int x, y;
    scanf("%d%d", &x, &y); // Read button values of current controller
    
    if (x == y) {
      // If both buttons have same value, only way to win is if #+ = #- 
      puts(2 * f == n ? "YES" : "NO");
      continue;
    }
    
    if (x > y)
      swap(x, y); // Ensure x <= y for easier computation
    
    // Key check:
    // Let k be number of times we press smaller button x
    // Then we must press larger button y exactly (f - k) times such that:
    // k * x + (f - k) * y = 0  =>  k = (f * y) / (y - x)
    
    ll mid = 1ll * x * (n - 2 * f); // Compute numerator for k = (f * y) / (y - x)
    if (mid % (y - x) == 0 && mid / (y - x) <= f) // Check if k is integer and in [0, f]
      puts("YES");
    else
      puts("NO");
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByTier/L/1776_L_Controllers.py
# Problem: CF 1776 L - Controllers
# https://codeforces.com/contest/1776/problem/L

"""
Problem: Determine if it's possible to achieve a score of 0 in a game with n rounds,
where each round has either '+' or '-' symbol and player chooses between two buttons
with values a and b. The goal is to check whether a given controller (a, b) allows
the player to win the game.

Approach:
- Count number of '+' and '-' symbols in the string.
- For a given controller with button values a and b:
  - If a == b:
    - Check if number of '+'s equals number of '-'s (p == m).
  - Else:
    - Compute score as p*a - m*b.
    - Check if there exists a valid assignment of buttons to rounds such that
      final score is 0. Using algebraic derivation, we check if certain modular
      conditions hold.

Time Complexity: O(n + q), where n is the length of string s and q is number of queries.
Space Complexity: O(1) additional space excluding input storage.

Algorithms/Techniques:
- Linear time counting
- Modular arithmetic and algebraic analysis
"""

import sys

input = sys.stdin.readline
n = int(input())
p = input().count("+")  # Count of '+' signs
m = n - p               # Count of '-' signs

for _ in range(int(input())):
    a, b = map(int, input().split())
    if not (min(a, b) * p - max(a, b) * m <= 0 <= max(a, b) * p - min(a, b) * m):
        print("NO")
        continue
    if a == b:
        print("YES" if p == m else "NO")  # Only possible if equal number of + and - signs
    else:
        print("YES" if (m * b - p * b) % (b - a) == 0 else "NO")  # Check modular condition for valid assignment


# https://github.com/VaHiX/codeForces/